tidy3d/plugins/adjoint/components/geometry.py

```diff
@@ -0,0 +1,189 @@
+"""Defines jax-compatible geometries and their conversion to grad monitors."""
+from __future__ import annotations
+
+from typing import Tuple, Union
+
+import pydantic as pd
+import numpy as np
+import jax.numpy as jnp
+from jax.tree_util import register_pytree_node_class
+
+from ....components.types import Bound
+from ....components.geometry import Box
+from ....components.data.monitor_data import FieldData, PermittivityData
+from ....components.monitor import FieldMonitor, PermittivityMonitor
+from ....constants import fp_eps
+
+from .base import JaxObject
+from .types import JaxFloat, validate_jax_tuple
+
+# number of cells along each dimension of the surface area to be integrated over
+NUM_CELLS_DIM = 401
```

Comment:
How is this number determined?

Alternatively, can we integrate over the grids projected to the box surface directly? This can be more accurate e.g. when nonuniform mesh is applied.

Reply:
it was determined somewhat empirically as a high number that wasn't too high to make things slow.. There are some complications involving using the fields at the grid points. The main reason is that with subpixel on, we make the assumption that the FDTD simulation is accurately modeling the `Structure.geometry`. So we want to integrate over the Box surface areas using the Box.bounds as the limits of integration. It is a lot simpler to do this integral assuming evenly spaced integration points and using trapezoidal rule and interpolation. If we use the raw grid points, it's still possible, but adds a bit of complication determining the dL at the edges and the interpolation approach seems to give very accurate gradients already. Does this make sense?

Reply:
This does seem a bit weird though. I think it should rather be a number of cells per wavelength (in the material?) kinda thing? Currently, you'd take the same number of cells in a very small and a very large surface? Which is not great.

Reply:
It is a bit weird, the intention was to pick a number of integration points that is never too large to crash the program due to running out of memory when interpolating.  However, probably a min cells per wavelength in material makes sense, maybe with an upper limit on number of cells for robustness. I can make that change. However, note that I don't observe this part of the adjoint pipeline slowing things down at all, so I don't expect to see any speed improvements or anything.

Reply:
My concern is not so much speed but accuracy if you have too few points on a surface that's e.g. much larger than what you've tested so far.

Reply:
Alright, I just changed it to support instead a `PTS_PER_WVL_INTEGRATION = 20` variable and now any surface integration over a structure will first compute the wavelength inside the structure and use this, `PTS_PER_WVL_INTEGRATION`, and the length of the edge to determine how many integration points to use along that dimension.

Reply:
Similar modifications to here https://github.com/flexcompute/tidy3d/blob/904392107d098a160168c4ad90cdf10743a69610/tidy3d/plugins/adjoint/components/medium.py#L60-L62 ?